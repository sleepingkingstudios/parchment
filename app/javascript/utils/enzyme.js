import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
