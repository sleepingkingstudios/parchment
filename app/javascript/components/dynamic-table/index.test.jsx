import React from 'react';
import { shallow } from 'enzyme';

import DynamicTable from './index';

describe('<DynamicTable />', () => {
  const columns = [
    { label: true, prop: 'id' },
    { label: true, prop: 'name' },
    { label: true, prop: 'price' },
    { label: true, prop: 'quantity' },
  ];
  const message = 'There are no widgets matching the criteria.';
  const defaultProps = { columns, data: [], message };

  describe('with cellProps: value', () => {
    const cellProps = { key: 'value' };
    const data = [
      {
        id: '00000000-0000-0000-0000-000000000000',
        name: 'Self-sealing Stem Bolt',
        price: '10.00',
        quantity: 50,
      },
    ];

    it('should render the table rows', () => {
      const rendered = shallow(
        <DynamicTable {...defaultProps} data={data} cellProps={cellProps} />,
      );
      const rows = rendered.find('DynamicTableRow');

      expect(rows.length).toEqual(data.length);

      data.forEach((item, index) => {
        const row = rows.at(index);

        expect(row).toHaveProp({ cellProps });
        expect(row).toHaveProp({ columns });
        expect(row).toHaveProp({ data: item });
      });
    });
  });

  describe('with className: value', () => {
    const className = 'custom-table';

    it('should render the dynamic table', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} className={className} />);

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table');
      expect(rendered).toHaveClassName('container');
      expect(rendered).toHaveClassName(className);
    });
  });

  describe('with data: an empty array', () => {
    const data = [];

    it('should render the dynamic table', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table');
      expect(rendered).toHaveClassName('container');
    });

    it('should render the table header', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);
      const header = rendered.find('DynamicTableHeader');

      expect(header).toExist();
      expect(header).toHaveProp({ columns });
    });

    it('should render the empty message', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);
      const emptyMessage = rendered.find('DynamicTableEmptyMessage');

      expect(emptyMessage).toExist();
      expect(emptyMessage).toHaveProp({ message });
    });

    it('should not render any rows', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);
      const rows = rendered.find('DynamicTableRow');

      expect(rows.length).toEqual(0);
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with data: an array of objects', () => {
    const data = [
      {
        id: '00000000-0000-0000-0000-000000000000',
        name: 'Self-sealing Stem Bolt',
        price: '10.00',
        quantity: 50,
      },
      {
        id: '00000000-0000-0000-0000-000000000001',
        name: 'Inverse Spatial Manifold',
        price: '1000.00',
        quantity: 10,
      },
      {
        id: '00000000-0000-0000-0000-000000000002',
        name: 'Ambrosia Software License',
        price: '10000.00',
        quantity: 5,
      },
    ];

    it('should render the dynamic table', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table');
      expect(rendered).toHaveClassName('container');
    });

    it('should render the table header', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);
      const header = rendered.find('DynamicTableHeader');

      expect(header).toExist();
      expect(header).toHaveProp({ columns });
    });

    it('should not render an empty message', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);
      const emptyMessage = rendered.find('DynamicTableEmptyMessage');

      expect(emptyMessage).not.toExist();
    });

    it('should render the table rows', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);
      const rows = rendered.find('DynamicTableRow');

      expect(rows.length).toEqual(data.length);

      data.forEach((item, index) => {
        const row = rows.at(index);

        expect(row).toHaveProp({ columns });
        expect(row).toHaveProp({ data: item });
      });
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<DynamicTable {...defaultProps} data={data} />);

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with resourceName: value', () => {
    const data = [
      {
        id: '00000000-0000-0000-0000-000000000000',
        name: 'Self-sealing Stem Bolt',
        price: '10.00',
        quantity: 50,
      },
    ];
    const resourceName = 'widgets';

    it('should render the dynamic table', () => {
      const rendered = shallow(
        <DynamicTable {...defaultProps} data={data} resourceName={resourceName} />,
      );

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table');
      expect(rendered).toHaveClassName('widgets-table');
      expect(rendered).toHaveClassName('container');
    });

    it('should render the table header', () => {
      const rendered = shallow(
        <DynamicTable {...defaultProps} data={data} resourceName={resourceName} />,
      );
      const header = rendered.find('DynamicTableHeader');

      expect(header).toExist();
      expect(header).toHaveProp({ columns });
      expect(header).toHaveProp({ resourceName });
    });

    it('should render the table rows', () => {
      const rendered = shallow(
        <DynamicTable {...defaultProps} data={data} resourceName={resourceName} />,
      );
      const rows = rendered.find('DynamicTableRow');

      expect(rows.length).toEqual(data.length);

      data.forEach((item, index) => {
        const row = rows.at(index);

        expect(row).toHaveProp({ columns });
        expect(row).toHaveProp({ data: item });
        expect(row).toHaveProp({ resourceName });
      });
    });
  });
});
