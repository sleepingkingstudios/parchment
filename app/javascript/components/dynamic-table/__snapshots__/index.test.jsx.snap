// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicTable /> with data: an array of objects should match the snapshot 1`] = `
<div
  className="dynamic-table container"
>
  <DynamicTableHeader
    columns={
      Array [
        Object {
          "label": true,
          "prop": "id",
        },
        Object {
          "label": true,
          "prop": "name",
        },
        Object {
          "label": true,
          "prop": "price",
        },
        Object {
          "label": true,
          "prop": "quantity",
        },
      ]
    }
    resourceName={null}
  />
  <DynamicTableRow
    cellProps={Object {}}
    columns={
      Array [
        Object {
          "label": true,
          "prop": "id",
        },
        Object {
          "label": true,
          "prop": "name",
        },
        Object {
          "label": true,
          "prop": "price",
        },
        Object {
          "label": true,
          "prop": "quantity",
        },
      ]
    }
    data={
      Object {
        "id": "00000000-0000-0000-0000-000000000000",
        "name": "Self-sealing Stem Bolt",
        "price": "10.00",
        "quantity": 50,
      }
    }
    key="00000000-0000-0000-0000-000000000000"
    resourceName={null}
  />
  <DynamicTableRow
    cellProps={Object {}}
    columns={
      Array [
        Object {
          "label": true,
          "prop": "id",
        },
        Object {
          "label": true,
          "prop": "name",
        },
        Object {
          "label": true,
          "prop": "price",
        },
        Object {
          "label": true,
          "prop": "quantity",
        },
      ]
    }
    data={
      Object {
        "id": "00000000-0000-0000-0000-000000000001",
        "name": "Inverse Spatial Manifold",
        "price": "1000.00",
        "quantity": 10,
      }
    }
    key="00000000-0000-0000-0000-000000000001"
    resourceName={null}
  />
  <DynamicTableRow
    cellProps={Object {}}
    columns={
      Array [
        Object {
          "label": true,
          "prop": "id",
        },
        Object {
          "label": true,
          "prop": "name",
        },
        Object {
          "label": true,
          "prop": "price",
        },
        Object {
          "label": true,
          "prop": "quantity",
        },
      ]
    }
    data={
      Object {
        "id": "00000000-0000-0000-0000-000000000002",
        "name": "Ambrosia Software License",
        "price": "10000.00",
        "quantity": 5,
      }
    }
    key="00000000-0000-0000-0000-000000000002"
    resourceName={null}
  />
</div>
`;

exports[`<DynamicTable /> with data: an empty array should match the snapshot 1`] = `
<div
  className="dynamic-table container"
>
  <DynamicTableHeader
    columns={
      Array [
        Object {
          "label": true,
          "prop": "id",
        },
        Object {
          "label": true,
          "prop": "name",
        },
        Object {
          "label": true,
          "prop": "price",
        },
        Object {
          "label": true,
          "prop": "quantity",
        },
      ]
    }
    resourceName={null}
  />
  <DynamicTableEmptyMessage
    message="There are no widgets matching the criteria."
  />
</div>
`;
